testonly/mdm/mdm.go

```diff
@@ -0,0 +1,163 @@
+	"github.com/google/trillian/client"
+	"github.com/google/trillian/monitoring"
+	"github.com/google/trillian/types"
+)
+
+// MergeDelayMonitor submits leaves to a Log and measures merge delay.
+type MergeDelayMonitor struct {
+	client []*client.LogClient
+	logID  int64
+	opts   MergeDelayOptions
+}
+
+// MergeDelayOptions holds the parameters for a MergeDelayMonitor.
+type MergeDelayOptions struct {
+	ParallelAdds  int
+	LeafSize      int
+	NewLeafChance int // percentage
+	EmitInterval  time.Duration
+	MetricFactory monitoring.MetricFactory
+}
+
+// NewMonitor creates a MergeDelayMonitor instance for the given log ID, accessed
+// via the cl client.
+func NewMonitor(ctx context.Context, logID int64, cl trillian.TrillianLogClient, acl trillian.TrillianAdminClient, opts MergeDelayOptions) (*MergeDelayMonitor, error) {
+	if opts.MetricFactory == nil {
+		opts.MetricFactory = monitoring.InertMetricFactory{}
+	}
+	if opts.EmitInterval == 0 {
+		opts.EmitInterval = 10 * time.Second
+	}
+	metricsOnce.Do(func() { initMetrics(opts.MetricFactory) })
+
+	tree, err := acl.GetTree(ctx, &trillian.GetTreeRequest{TreeId: logID})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get tree %d: %v", logID, err)
+	}
+	verifier, err := client.NewLogVerifierFromTree(tree)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build verifier: %v", err)
+	}
+	clients := make([]*client.LogClient, opts.ParallelAdds)
+	for i := 0; i < opts.ParallelAdds; i++ {
+		clients[i] = client.New(logID, cl, verifier, types.LogRootV1{})
+	}
+
+	mdm := MergeDelayMonitor{logID: logID, opts: opts, client: clients}
+	return &mdm, nil
+}
+
+// Monitor runs merge delay monitoring until its context is cancelled or an error occurs.
+func (m *MergeDelayMonitor) Monitor(ctx context.Context) error {
+	glog.Infof("starting %d parallel monitor instances", m.opts.ParallelAdds)
+	errs := make(chan error, m.opts.ParallelAdds)
+	var wg sync.WaitGroup
+	wg.Add(m.opts.ParallelAdds)
+	for i := 0; i < m.opts.ParallelAdds; i++ {
+		go func(i int) {
+			defer wg.Done()
+			if err := m.monitor(ctx, i); err != nil {
+				errs <- err
+			}
+		}(i)
+	}
+
+	ticker := time.NewTicker(m.opts.EmitInterval)
+	defer ticker.Stop()
+	go func(c <-chan time.Time) {
+		for range c {
+			countT, totalT := m.Stats(true)
+			if countT > 0 {
+				glog.Infof("new leaves: %d in %f secs, average %f secs", countT, totalT, totalT/float64(countT))
+			}
+			countF, totalF := m.Stats(false)
+			if countF > 0 {
+				glog.Infof("dup leaves: %d in %f secs, average %f secs", countF, totalF, totalF/float64(countF))
+			}
+		}
+	}(ticker.C)
+
+	wg.Wait()
+	close(errs)
+	var lastErr error
+	for err := range errs {
+		glog.Errorf("AddLeaf failure: %v", err)
+		lastErr = err
+	}
+	return lastErr
+}
+
+func (m *MergeDelayMonitor) monitor(ctx context.Context, idx int) error {
+	logIDLabel := strconv.FormatInt(m.logID, 10)
+	data := make([]byte, m.opts.LeafSize)
+	createNew := true
+	for {
+		if rand.Intn(100) < m.opts.NewLeafChance {
```

Comment:
Could maybe pull this out into a closure that returns a new leaf with a probability and clear up the createNew at the top and bottom of the loop. Just an idea, might not be a good idea.

Reply:
The initial `createNew := true` is a wart that's needed to ensure that the first iteration always creates a new leaf (as there's no old leaf to re-use) &ndash; added a comment to help clarity.

Reply:
Yeah I thought this could be a func() that would wrap up the createNew and the data slice. Not something I'm going to insist on.